File: scripts/explore-owid/list-charts.py
"""
list-charts.py
Fetches all charts from the OWID Search API and writes them to charts.json

Usage: python list-charts.py [--output charts.json]

API docs: https://docs.owid.io/projects/etl/api/search-api/
"""

import argparse
import json
import sys
import urllib.request

BASE_URL = "https://ourworldindata.org/api/search"
HITS_PER_PAGE = 100


def fetch_page(page: int) -> dict:
    url = f"{BASE_URL}?type=charts&hitsPerPage={HITS_PER_PAGE}&page={page}"
    req = urllib.request.Request(
        url,
        headers={"User-Agent": "Mozilla/5.0 (compatible; owid-chart-explorer/1.0)"},
    )
    with urllib.request.urlopen(req) as res:
        return json.loads(res.read())


def fetch_all_charts() -> list:
    print("Fetching page 0 to get total count...")
    first = fetch_page(0)

    total_hits = first.get("nbHits", 0)
    total_pages = first.get("nbPages", (total_hits + HITS_PER_PAGE - 1) // HITS_PER_PAGE)
    print(f"Total charts: {total_hits}, pages: {total_pages}")

    all_charts = list(first.get("results", []))

    for page in range(1, total_pages):
        print(f"\rFetching page {page + 1}/{total_pages}...", end="", flush=True)
        data = fetch_page(page)
        all_charts.extend(data.get("results", []))

    print(f"\nDone. Fetched {len(all_charts)} charts.")
    return all_charts


def main():
    parser = argparse.ArgumentParser(description="List all OWID charts")
    parser.add_argument("--output", default="charts.json", help="Output file path")
    args = parser.parse_args()

    charts = fetch_all_charts()

    # Print a sample to stdout
    print("\nSample (first 20):")
    print("─" * 80)
    for c in charts[:20]:
        title = (c.get("title") or "")[:60].ljust(60)
        slug = c.get("slug") or ""
        print(f"{title}  {slug}")
    print("─" * 80)

    print(f"\nWriting all {len(charts)} charts to {args.output}...")
    with open(args.output, "w") as f:
        json.dump(charts, f, indent=2)
    print("Done.")


if __name__ == "__main__":
    main()

File: scripts/explore-owid/list-charts.js
/**
 * list-charts.js
 * Fetches all charts from the OWID Search API and writes them to charts.json
 *
 * Usage: node list-charts.js [--output charts.json]
 *
 * API docs: https://docs.owid.io/projects/etl/api/search-api/
 */

const BASE_URL = "https://ourworldindata.org/api/search";
const HITS_PER_PAGE = 100;

async function fetchPage(page) {
  const url = `${BASE_URL}?type=charts&hitsPerPage=${HITS_PER_PAGE}&page=${page}`;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`HTTP ${res.status} on page ${page}`);
  return res.json();
}

async function fetchAllCharts() {
  console.log("Fetching page 0 to get total count...");
  const first = await fetchPage(0);

  const totalPages = first.nbPages ?? Math.ceil(first.nbHits / HITS_PER_PAGE);
  const totalHits = first.nbHits;
  console.log(`Total charts: ${totalHits}, pages: ${totalPages}`);

  const allCharts = [...first.results];

  for (let page = 1; page < totalPages; page++) {
    process.stdout.write(`\rFetching page ${page + 1}/${totalPages}...`);
    const data = await fetchPage(page);
    allCharts.push(...data.results);
  }

  console.log(`\nDone. Fetched ${allCharts.length} charts.`);
  return allCharts;
}

async function main() {
  const outputArg = process.argv.indexOf("--output");
  const outputFile = outputArg !== -1 ? process.argv[outputArg + 1] : "charts.json";

  const charts = await fetchAllCharts();

  // Print a summary table to stdout
  console.log("\nSample (first 20):");
  console.log("─".repeat(80));
  for (const c of charts.slice(0, 20)) {
    const title = (c.title ?? "").slice(0, 60).padEnd(60);
    const slug = c.slug ?? "";
    console.log(`${title}  ${slug}`);
  }
  console.log("─".repeat(80));
  console.log(`\nWriting all ${charts.length} charts to ${outputFile}...`);

  const fs = await import("fs");
  fs.writeFileSync(outputFile, JSON.stringify(charts, null, 2));
  console.log("Done.");
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
